8329088: Stack chunk thawing races with concurrent GC stack iteration

fisk opened the pull request:
When we thaw the last frame from a stack chunk, we non-atomically set the stack pointer (sp), and set its argsize to 0. Unfortunately, GC threads may iterate over the frames of the stack chunk concurrently. When initializing their stack frame iterator, they read the sp and argsize racingly. Since there is no synchronization between the threads, we may observe inconsistent pairs of sp and argsize, for example the updated sp with a stale argsize, or the updated argsize with a stale sp.

At the core of the problem, the stack chunks define sp and argsize. The argsize is used to calculate where the bottom of the stack chunk is, which is required to determine if it is empty or not. This patch proposes to switch things around and store the bottom directly in the chunk, instead of argsize. Instead, argsize is calculated from the bottom. By changing the relationship of which property is stored and which property is calculated, we can simplify this code quite a bit.

In the new model, is_empty() is true iff sp and bottom are exactly the same. Bottom is only set during freezing, never during thawing. The bottom is initialized whenever the bottom frame is frozen, and left untouched during thawing. Unlike thawing, the freeze operation does not race with the GC by design. Hence we have moved one of the racy mutations to the operation that doesn't race with the GC. The GC is now only exposed to changing sp(). It doesn't matter if it observes the old or new sp(), now that we have removed the only source if inconsistency describing said frame (racing argsize).

Testing: tier1-5, manual testing of test/jdk/jdk/internal/vm/Continuation

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329088](https://bugs.openjdk.org/browse/JDK-8329088): Stack chunk thawing races with concurrent GC stack iteration (**Bug** - P2)


### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**) ⚠️ Review applies to [40ea7943](https://git.openjdk.org/jdk/pull/18643/files/40ea7943097e489a349d69977c2b065b26a73f70)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18643/head:pull/18643` \
`$ git checkout pull/18643`

Update a local copy of the PR: \
`$ git checkout pull/18643` \
`$ git pull https://git.openjdk.org/jdk.git pull/18643/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18643`

View PR using the GUI difftool: \
`$ git pr show -t 18643`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18643.diff">https://git.openjdk.org/jdk/pull/18643.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18643#issuecomment-2039021559)

fisk commented:
/label remove core-libs

stefank commented on src/hotspot/share/oops/stackChunkOop.inline.hpp:
```diff
@@ -108,7 +108,14 @@ inline void stackChunkOopDesc::set_cont_raw(oop value)    { jdk_internal_vm_Stac
-inline int stackChunkOopDesc::bottom() const { return stack_size() - argsize() - frame::metadata_words_at_top; }
+inline int stackChunkOopDesc::argsize() const {
+  assert(!is_empty(), "should not ask for argsize in empty chunk");
+  if (is_empty()) {
+    return 0;
+  } else {
```
Should this be removed?

stefank commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -653,13 +645,11 @@ bool FreezeBase::freeze_fast_new_chunk(stackChunkOop chunk) {
   const int chunk_start_sp = cont_size() + frame::metadata_words;
-  assert(chunk_start_sp == chunk->stack_size(), "");
+
+  chunk->set_max_thawing_size(cont_size());
```
Should this move be reverted?

stefank commented on src/hotspot/share/runtime/continuationJavaClasses.hpp:
```diff
@@ -107,13 +107,15 @@ class jdk_internal_vm_StackChunk: AllStatic {
   static inline int size(oop chunk);
   static inline void set_size(HeapWord* chunk, int value);
 
+  static inline void set_bottom(HeapWord* chunk, int value);
```
Shouldn't this be moved down to the other set_bottom? The two set_sp functions are held together.

stefank reviewed (approved):
Looks good. There's a few nits that could be worth considering.

fisk commented:
Thanks for the review @stefank. I fixed the nits you found.

shipilev commented on src/hotspot/share/oops/oop.inline.hpp:
```diff
@@ -238,6 +238,8 @@ inline void   oopDesc::short_field_put(int offset, jshort value)    { *field_add
 
 inline jint oopDesc::int_field(int offset) const                    { return *field_addr<jint>(offset);     }
 inline void oopDesc::int_field_put(int offset, jint value)          { *field_addr<jint>(offset) = value;    }
+inline jint oopDesc::int_field_relaxed(int offset) const            { return Atomic::load(field_addr<jint>(offset)); }
+inline void oopDesc::int_field_put_relaxed(int offset, jint value)  { Atomic::store(field_addr<jint>(offset), value); }
```
I have a stylistic question/suggestion. These are basically Java heap accessors, shouldn't they go through `HeapAccess<MO_RELAXED>::{load,store}`? This would also match the style already used in this file.

stefank replied:
We don't use HeapAccess to access primitive values in objects. HeapAccess is only used when we access oops.

We do use RawAccess in some of these functions though, but we do that because there's no support for MQ_SEQ_CST in the Atomic APIs.

fisk replied:
Ideally the SEQ_CST support would live in Atomic where it arguably belongs. Then we could remove all primitive support from the Access API and have a clear distinction that it's an oop thing.

shipilev reviewed (commented):
This looks reasonable, I have a question, though:

fisk commented:
> This looks reasonable, I have a question, though:

Thanks for the review!

pchilano commented:
> In the new model, is_empty() is true iff sp and bottom are exactly the same. Bottom is only set during freezing, never during thawing. The bottom is initialized whenever the bottom frame is frozen, and left untouched during thawing. Unlike thawing, the freeze operation does not race with the GC by design. Hence we have moved one of the racy mutations to the operation that doesn't race with the GC. The GC is now only exposed to changing sp(). It doesn't matter if it observes the old or new sp(), now that we have removed the only source if inconsistency describing said frame (racing argsize).
>
So if the race happens only when resetting the stackChunk values when thawing the last frame, wouldn't it be enough to avoid clearing the argsize there? Because if we read the new sp when creating the stack frame iterator, regardless of the argsize value read, is_done() will be true so we won't iterate any frame. I'm trying to understand if the new model is needed to fix the race or that is part of a cleanup/refactoring.

pchilano commented:
> Unlike thawing, the freeze operation does not race with the GC by design.
>
Is this with the changes in the allocation code in this patch or even before those there was no race?

fisk commented:
> > In the new model, is_empty() is true iff sp and bottom are exactly the same. Bottom is only set during freezing, never during thawing. The bottom is initialized whenever the bottom frame is frozen, and left untouched during thawing. Unlike thawing, the freeze operation does not race with the GC by design. Hence we have moved one of the racy mutations to the operation that doesn't race with the GC. The GC is now only exposed to changing sp(). It doesn't matter if it observes the old or new sp(), now that we have removed the only source if inconsistency describing said frame (racing argsize).
> 
> So if the race happens only when resetting the stackChunk values when thawing the last frame, wouldn't it be enough to avoid clearing the argsize there? Because if we read the new sp when creating the stack frame iterator, regardless of the argsize value read, is_done() will be true so we won't iterate any frame. I'm trying to understand if the new model is needed to fix the race or that is part of a cleanup/refactoring.

I thought about going in that direction as well. But in the end, I found it to be more direct and easy to understand if the property being tracked directly in the stack chunk is bottom, and argsize is the computed property, as opposed to the other way around.

pchilano commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -572,7 +564,7 @@ int FreezeBase::size_if_fast_freeze_available() {
   // Consider leaving the chunk's argsize set when emptying it and removing the following branch,
   // although that would require changing stackChunkOopDesc::is_empty
-  if (chunk_sp < chunk->stack_size()) {
+  if (!chunk->is_empty()) {
```
Seems you have implemented the suggestion in the comment so we can remove this branch and unconditionally decrement total_size_needed.

fisk replied:
I currently have an assert that checks that you shouldn't be asking for the argsize() if the chunk is empty, because it is so error prone. I think I'd like to keep the assert though - it was quite useful.

pchilano replied:
We should be okay since _cont.argsize gets it from the ContinuationEntry. I tested it a bit and we would also need to update _fast_freeze_size to be cont_size() in the chunk empty case before calling freeze_fast_copy() otherwise we hit an assert there. But I can do this in another RFR if you want.

fisk replied:
That would be great, thank you!

pchilano commented on src/hotspot/share/oops/stackChunkOop.inline.hpp:
```diff
@@ -132,10 +135,7 @@ inline intptr_t* stackChunkOopDesc::from_offset(int offset) const {
 }
 
 inline bool stackChunkOopDesc::is_empty() const {
-  assert(sp() <= stack_size(), "");
```
Maybe keep assert(sp() <= bottom(), "")?

fisk replied:
Fixed!

pchilano reviewed (commented):
Thanks, looks good to me. Only a few comments.

fisk commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -630,13 +621,14 @@ void FreezeBase::freeze_fast_existing_chunk() {
     chunk->set_max_thawing_size(cont_size());
-    chunk->set_argsize(_cont.argsize());
+    chunk->set_bottom(chunk_start_sp - _cont.argsize());
+    chunk->set_sp(chunk->bottom());
```
We used to set sp to stack_size() before to say the stack is_empty during thaw. It had an imprecise and fuzzy notion of being empty that accepted various different bottoms depending on argsize. The new definition of is_empty is that bottom == sp, precisely. I don't want to update bottom during thaw, only when the chunk is initialized. Now that we are re-initializing the chunk with a new bottom (due to different arg size), we have to set the sp to bottom to signify that the chunk is_empty() right now.

fisk commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -654,12 +646,10 @@ bool FreezeBase::freeze_fast_new_chunk(stackChunkOop chunk) {
   // in a fresh chunk, we freeze *with* the bottom-most frame's stack arguments.
   // They'll then be stored twice: in the chunk and in the parent chunk's top frame
   const int chunk_start_sp = cont_size() + frame::metadata_words;
-  assert(chunk_start_sp == chunk->stack_size(), "");
```
It is. Fixed!

pchilano commented on src/hotspot/share/runtime/continuationFreezeThaw.cpp:
```diff
@@ -630,13 +621,14 @@ void FreezeBase::freeze_fast_existing_chunk() {
   } else { // the chunk is empty
-    DEBUG_ONLY(_empty = true;)
-    const int chunk_start_sp = chunk->sp();
+    const int chunk_start_sp = chunk->stack_size() - frame::metadata_words_at_top;
```
Do we need the minus frame::metadata_words_at_top? Since the chunk is empty I would expect the chunk_start_sp to be the same as for the new chunk case which is just chunk->stack_size().

fisk replied:
I think you are right; I changed it to chunk->stack_size() as you propose, and adjusted the bottom to remove the metadata part instead. I think this looks nicer.

pchilano reviewed (approved):
Thanks Erik, looks good to me. Some extra comments.

pchilano reviewed (approved):
Thanks Erik!

txominpelu commented:
Hello :wave:, I wonder @fisk if you were able to reproduce the situation where a GC thread sees inconsistent values for `sp` and `argsize` and whether it would make sense to write a test forcing that situation to happen to prove that this PR fixes the issue and to avoid regressions ? I'm not super experienced with the hotspot code, but I'd be happy to try to write a test for that if you think that it makes sense.

fisk commented:
> Hello 👋, I wonder @fisk if you were able to reproduce the situation where a GC thread sees inconsistent values for `sp` and `argsize` and whether it would make sense to write a test forcing that situation to happen to prove that this PR fixes the issue and to avoid regressions ? I'm not super experienced with the hotspot code, but I'd be happy to try to write a test for that if you think that it makes sense.

Writing a test for this would be rather tricky, as you have to put the GC in a rather particular state, to have it actually race with thawing. If you want to give it a shot, then that would be nice, but it won't be easy I think.

fisk commented:
/integrate
